gax-java/gax/src/test/java/com/google/api/gax/tracing/MetricsTracerTest.java

```diff
@@ -0,0 +1,247 @@
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testCancelledExample() {
+    // initialize mock-request
+    Object mockCancelledRequest = new Object();
+
+    // Attempt #1
+    metricsTracer.attemptStarted(mockCancelledRequest, 0);
+    metricsTracer.attemptCancelled();
+    metricsTracer.operationCancelled();
+
+    long count = 1;
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "CANCELLED",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(count, attributes);
+    verify(metricsRecorder).recordOperationCount(count, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+    verify(metricsRecorder).recordOperationLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptFailedRetriesExhaustedExample() {
+    // initialize mock-request
+    Object mockRequest = new Object();
+
+    // Attempt #1
+    metricsTracer.attemptStarted(mockRequest, 0);
+    ApiException error0 =
+        new DeadlineExceededException(
+            "deadline exceeded", null, new FakeStatusCode(Code.DEADLINE_EXCEEDED), false);
+
+    metricsTracer.attemptFailedRetriesExhausted(error0);
+    metricsTracer.operationFailed(error0);
+
+    long count = 1;
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "DEADLINE_EXCEEDED",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(count, attributes);
+    verify(metricsRecorder).recordOperationCount(count, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+    verify(metricsRecorder).recordOperationLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testAttemptPermanentFailureExample() {
+    // initialize mock-request
+    Object mockRequest = new Object();
+
+    // Attempt #1
+    metricsTracer.attemptStarted(mockRequest, 0);
+
+    ApiException error0 =
+        new NotFoundException("not found", null, new FakeStatusCode(Code.NOT_FOUND), false);
+
+    metricsTracer.attemptFailedRetriesExhausted(error0);
+    metricsTracer.operationFailed(error0);
+
+    long count = 1;
+    Map<String, String> attributes =
+        ImmutableMap.of(
+            "status", "NOT_FOUND",
+            "method_name", "fake_service.fake_method");
+
+    verify(metricsRecorder).recordAttemptCount(count, attributes);
+    verify(metricsRecorder).recordOperationCount(count, attributes);
+    verify(metricsRecorder).recordAttemptLatency(anyDouble(), eq(attributes));
+    verify(metricsRecorder).recordOperationLatency(anyDouble(), eq(attributes));
+
+    verifyNoMoreInteractions(metricsRecorder);
+  }
+
+  @Test
+  public void testErrorConversion() {
```

Comment:
We need more test cases for `extractStatus` method.

Reply:
Added more cases.